File: pa2/Lex.c
/****************************************************************************************
* Guangyang Chen
* 1518931
* PA2
* Lex.c
* Client program
*****************************************************************************************/
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"
#define MAX_LEN 1000

int main(int argc, char * argv[]){
   
   FILE *in, *in2, *out;
   char line[MAX_LEN];
   int count = 0;
   int i = 0;
   
   //check 3 arugments 
   if( argc != 3 ){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   //open files to read and write 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      printf( "Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out==NULL ){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }
   
   //count the lines
   while( fgets(line, MAX_LEN, in) != NULL) {
      count++;
   }
   fclose(in);
   
   in2 = fopen(argv[1], "r");
   char *str[count];
   while( fgets(line, MAX_LEN, in2) != NULL) {
      str[i] = malloc(strlen(line)+1);
      strcpy(str[i], line);
      i++;
   }
   
   // Create new List ADT
   List list = newList();

	append(list,0);
   moveFront(list);

   // Insertion Sort 
   for(int j = 1; j < count; ++j) {
      char *temp = str[j];
      moveBack(list);
      while(index(list) >= 0 && strcmp(temp, str[get(list)]) < 0){
         movePrev(list); 
      }
      if(index(list) >= 0){
         insertAfter(list,j);
      } else {
         prepend(list,j);
      }
   }
   moveFront(list);

   while(index(list)>=0) {
      fprintf(out, str[get(list)]); 
      moveNext(list);
   }
  
   // Free Array and list ADT
   for(i = 0; i < count; i++) {
      free(str[i]);
   }
   clear(list);
   freeList(&list);
     
   // Close files
   fclose(in2);
   fclose(out);
}

File: pa1/Lex.java
import java.io.*;
import java.util.Scanner;
public class Lex {

	public static void main(String[] args) throws IOException {
		Scanner in = null;
		PrintWriter out = null;
		String[] words = null;
		int i = 0;
		int n = 0;
		int lineNumber = 0;

		// check two command line arguments
		if (args.length < 2) {
			System.err.println("Usage: <infile> <outfile>");
			System.exit(1);
		}

		// open files
		try {
			// initialize input and output
			in = new Scanner(new File(args[0]));
			out = new PrintWriter(new FileWriter(args[1]));
		} catch (FileNotFoundException e) {
			System.err.println("Error: file target <" + args[0] + "> not found");
			System.exit(1);
		}
		
		// count lines
		while (in.hasNextLine()) {
			n++;
			in.nextLine();
		}
		// close file
		in.close();

		// Place words into Array
		words = new String[n];
		in = new Scanner(new File(args[0]));
		while( in.hasNextLine()){
			words[lineNumber++] = in.nextLine();
		}
		
		List L = new List();// Declare list for indexes
		int j,k; 
		String temp;
		L.append(0);//create list with zero_th element
		L.moveFront(); //set and define cursor
		
		for(j=1; j<words.length; j++){
			k = L.get(); //get pointer index
			i = j-1; //length
			temp = words[j];
			while(i>=0 && temp.compareTo(words[k])<0){
				i--; 
				L.movePrev(); 
				try{
					k = L.get();
				}catch(RuntimeException e){
					//get() has -1 index exception error
					break;
				}
			}
			if(L.index() >= 0){
				L.insertAfter(j); //insert after pointer cursor
			}else if(L.index() < 0){
				L.prepend(j); //put at the front of list  
			} 
			L.moveBack(); //reset cursor to the back of sorted list
		}
		
		//print
		L.moveFront();
		while(L.index()>=0){
			i = L.get();
			out.println(words[i]);
			L.moveNext();
		}
		//close file reading and writing
		in.close();
		out.close();
	}

}
